cmd/gaia/init/utils.go

```diff
@@ -0,0 +1,110 @@
+package init
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/cmd/gaia/app"
+	"github.com/cosmos/cosmos-sdk/codec"
+	amino "github.com/tendermint/go-amino"
+	cfg "github.com/tendermint/tendermint/config"
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/p2p"
+	"github.com/tendermint/tendermint/privval"
+	"github.com/tendermint/tendermint/types"
+)
+
+// WriteGenesisFile creates and writes the genesis configuration to disk. An
+// error is returned if building or writing the configuration to file fails.
+func WriteGenesisFile(
+	genFile, chainID string, validators []types.GenesisValidator, appState json.RawMessage,
+) error {
+
+	genDoc := types.GenesisDoc{
+		ChainID:    chainID,
+		Validators: validators,
+		AppState:   appState,
+	}
+
+	if err := genDoc.ValidateAndComplete(); err != nil {
+		return err
+	}
+
+	return genDoc.SaveAs(genFile)
+}
+
+// WriteGenesisFileWithTime creates and writes the genesis configuration to disk.
+// An error is returned if building or writing the configuration to file fails.
+func WriteGenesisFileWithTime(
+	genFile, chainID string, validators []types.GenesisValidator,
+	appState json.RawMessage, genTime time.Time,
+) error {
+
+	genDoc := types.GenesisDoc{
+		GenesisTime: genTime,
+		ChainID:     chainID,
+		Validators:  validators,
+		AppState:    appState,
+	}
+
+	if err := genDoc.ValidateAndComplete(); err != nil {
+		return err
+	}
+
+	return genDoc.SaveAs(genFile)
+}
+
+// read of create the private key file for this config
+func ReadOrCreatePrivValidator(privValFile string) crypto.PubKey {
+	var privValidator *privval.FilePV
+
+	if common.FileExists(privValFile) {
+		privValidator = privval.LoadFilePV(privValFile)
+	} else {
+		privValidator = privval.GenFilePV(privValFile)
+		privValidator.Save()
+	}
+
+	return privValidator.GetPubKey()
+}
+
+// InitializeNodeValidatorFiles creates private validator and p2p configuration files.
+func InitializeNodeValidatorFiles(
+	config *cfg.Config) (nodeID string, valPubKey crypto.PubKey, err error,
+) {
+
+	nodeKey, err := p2p.LoadOrGenNodeKey(config.NodeKeyFile())
+	if err != nil {
+		return
+	}
+
+	nodeID = string(nodeKey.ID())
+	valPubKey = ReadOrCreatePrivValidator(config.PrivValidatorFile())
+
+	return
+}
+
+func loadGenesisDoc(cdc *amino.Codec, genFile string) (genDoc types.GenesisDoc, err error) {
+	genContents, err := ioutil.ReadFile(genFile)
+	if err != nil {
+		return
+	}
+
+	err = cdc.UnmarshalJSON(genContents, &genDoc)
+	return
+}
+
+func initializeEmptyGenesis(
+	cdc *codec.Codec, genFile string, chainID string, overwrite bool,
+) (appState json.RawMessage, err error) {
```

Comment:
```
func initializeEmptyGenesis(cdc *codec.Codec, genFile, chainID string, overwrite bool) (
      appState json.RawMessage, err error) {
```

Reply:
^ also notice that `string` only needs to be written once

Reply:
ahhh yes, I'll update that 👍